src/util/pcp.cpp

```diff
@@ -0,0 +1,524 @@
+
+    // Make sure that we send from requested destination address, anything else will be
+    // rejected by a security-conscious router.
+    if (sock->Bind((struct sockaddr*)&bind_addr, bind_addrlen) != 0) {
+        LogPrintLevel(BCLog::NET, BCLog::Level::Warning, "pcp: Could not bind to address: %s\n", NetworkErrorString(WSAGetLastError()));
+        return MappingError::NETWORK_ERROR;
+    }
+
+    // Associate UDP socket to gateway.
+    if (sock->Connect((struct sockaddr*)&dest_addr, dest_addrlen) != 0) {
+        LogPrintLevel(BCLog::NET, BCLog::Level::Warning, "pcp: Could not connect to gateway: %s\n", NetworkErrorString(WSAGetLastError()));
+        return MappingError::NETWORK_ERROR;
+    }
+
+    // Use getsockname to get the address toward the default gateway (the internal address),
+    // in case we don't know what address to map
+    // (this is only needed if bind is INADDR_ANY, but it doesn't hurt as an extra check).
+    struct sockaddr_storage internal_addr;
+    socklen_t internal_addrlen = sizeof(struct sockaddr_storage);
+    if (sock->GetSockName((struct sockaddr*)&internal_addr, &internal_addrlen) != 0) {
+        LogPrintLevel(BCLog::NET, BCLog::Level::Warning, "pcp: Could not get sock name: %s\n", NetworkErrorString(WSAGetLastError()));
+        return MappingError::NETWORK_ERROR;
+    }
+    CService internal;
+    if (!internal.SetSockAddr((struct sockaddr*)&internal_addr)) return MappingError::NETWORK_ERROR;
+    LogPrintLevel(BCLog::NET, BCLog::Level::Debug, "pcp: Internal address after connect: %s\n", internal.ToStringAddr());
+
+    // Build request packet. Make sure the packet is zeroed so that reserved fields are zero
+    // as required by the spec (and not potentially leak data).
+    // Make sure there's space for the request header and MAP specific request data.
+    std::vector<uint8_t> request(PCP_HDR_SIZE + PCP_MAP_SIZE);
+    // Fill in request header, See RFC6887 Figure 2.
+    size_t ofs = 0;
+    request[ofs + PCP_HDR_VERSION_OFS] = PCP_VERSION;
+    request[ofs + PCP_HDR_OP_OFS] = PCP_REQUEST | PCP_OP_MAP;
+    WriteBE32(request.data() + ofs + PCP_HDR_LIFETIME_OFS, lifetime);
+    if (!PCPWrapAddress(Span(request).subspan(ofs + PCP_REQUEST_HDR_IP_OFS, ADDR_IPV6_SIZE), internal)) return MappingError::NETWORK_ERROR;
+
+    ofs += PCP_HDR_SIZE;
+
+    // Fill in MAP request packet, See RFC6887 Figure 9.
+    // Randomize mapping nonce (this is repeated in the response, to be able to
+    // correlate requests and responses, and used to authenticate changes to the mapping).
+    std::memcpy(request.data() + ofs + PCP_MAP_NONCE_OFS, nonce.data(), PCP_MAP_NONCE_SIZE);
+    request[ofs + PCP_MAP_PROTOCOL_OFS] = PCP_PROTOCOL_TCP;
+    WriteBE16(request.data() + ofs + PCP_MAP_INTERNAL_PORT_OFS, port);
+    WriteBE16(request.data() + ofs + PCP_MAP_EXTERNAL_PORT_OFS, port);
+    if (!PCPWrapAddress(Span(request).subspan(ofs + PCP_MAP_EXTERNAL_IP_OFS, ADDR_IPV6_SIZE), bind)) return MappingError::NETWORK_ERROR;
```

Comment:
This is intentional. For IPv4 we provide the all-zeroes address, because the router will use its own (NAT), for IPv6 we *want* the external address to be the same as the local address (NAT is not a thing, we want pinholing).

Reply:
Would all-zeros work as well for IPv6 / pinholing? If yes, then this can probably be simplified to always pass all-zeros.